worker/introspection/leases.go

```diff
@@ -187,3 +199,138 @@ func (h leaseHandler) format(leases *leases) map[string]interface{} {
 	}
 	return result
 }
+
+// ServeHTTP is part of the http.Handler interface.
+func (h *leaseHandler) revoke(w http.ResponseWriter, r *http.Request) {
+	if r.Method != http.MethodPost {
+		w.WriteHeader(http.StatusMethodNotAllowed)
+		fmt.Fprintln(w, "revoking a lease requires a POST request")
+		return
+	}
+
+	leaseKey, err := h.parseRevokeForm(r)
+	if err != nil {
+		w.WriteHeader(http.StatusBadRequest)
+		fmt.Fprintln(w, err.Error())
+		return
+	}
+
+	snapshot, err := h.snapshot()
+	if err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+		fmt.Fprintln(w, err.Error())
+		return
+	}
+
+	holder, ok := snapshot.Entries[leaseKey]
+	if !ok {
+		w.WriteHeader(http.StatusBadRequest)
+		if leaseKey.Namespace == lease.SingularControllerNamespace {
+			fmt.Fprintf(w, "singular lease for model %q not found\n", leaseKey.ModelUUID)
+		} else {
+			fmt.Fprintf(w, "application lease for model %q and app %q not found\n", leaseKey.ModelUUID, leaseKey.Lease)
+		}
+		return
+	}
+
+	if err := h.revokeLeadership(leaseKey, holder.Holder); err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+		fmt.Fprintln(w, err.Error())
+		return
+	}
+
+	if leaseKey.Namespace == lease.SingularControllerNamespace {
+		fmt.Fprintf(w, "singular lease for model %q revoked\n", leaseKey.ModelUUID)
+	} else {
+		fmt.Fprintf(w, "application lease for model %q and app %q revoked\n", leaseKey.ModelUUID, leaseKey.Lease)
+	}
+}
+
+func (h *leaseHandler) revokeLeadership(key raftlease.SnapshotKey, holder string) error {
+	command := &raftlease.Command{
+		Version:   raftlease.CommandVersion,
+		Operation: raftlease.OperationRevoke,
+		Namespace: key.Namespace,
+		ModelUUID: key.ModelUUID,
+		Lease:     key.Lease,
+		Holder:    holder,
+	}
+
+	bytes, err := command.Marshal()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	if h.runID == 0 {
+		source := rand.NewSource(h.clock.Now().UnixNano())
+		h.runID = rand.New(source).Int31()
+	}
+
+	// TODO(wallyworld): this should really be in a shared messages.go file under
+	// the top level pubsub package.
+	leaseRequestTopic := "lease.request"
```

Comment:
Can this become a const some where as to not have magic strings?

Reply:
that's what the todo was for :-)
but done now